drivers/ieee802154/ieee802154_dw1000.c

```diff
@@ -0,0 +1,1682 @@
+#include <init.h>
+#include <net/net_if.h>
+#include <net/net_pkt.h>
+
+#include <sys/byteorder.h>
+#include <string.h>
+#include <random/rand32.h>
+#include <debug/stack.h>
+#include <math.h>
+
+#include <drivers/gpio.h>
+#include <drivers/spi.h>
+
+#include <net/ieee802154_radio.h>
+#include "ieee802154_dw1000_regs.h"
+
+#define DT_DRV_COMPAT decawave_dw1000
+
+#define DWT_FCS_LENGTH			2U
+#define DWT_SPI_CSWAKEUP_FREQ		500000U
+#define DWT_SPI_SLOW_FREQ		2000000U
+#define DWT_SPI_TRANS_MAX_HDR_LEN	3
+#define DWT_SPI_TRANS_REG_MAX_RANGE	0x3F
+#define DWT_SPI_TRANS_SHORT_MAX_OFFSET	0x7F
+#define DWT_SPI_TRANS_WRITE_OP		BIT(7)
+#define DWT_SPI_TRANS_SUB_ADDR		BIT(6)
+#define DWT_SPI_TRANS_EXTEND_ADDR	BIT(7)
+
+#define DWT_TS_TIME_UNITS_FS		15650U /* DWT_TIME_UNITS in fs */
+
+#define DW1000_TX_ANT_DLY		16450
+#define DW1000_RX_ANT_DLY		16450
+
+/* SHR Symbol Duration in ns */
+#define UWB_PHY_TPSYM_PRF64		1017.63
+#define UWB_PHY_TPSYM_PRF16		993.59
+
+#define UWB_PHY_NUMOF_SYM_SHR_SFD	8
+
+/* PHR Symbol Duration Tdsym in ns */
+#define UWB_PHY_TDSYM_PHR_110K		8205.13
+#define UWB_PHY_TDSYM_PHR_850K		1025.64
+#define UWB_PHY_TDSYM_PHR_6M8		1025.64
+
+#define UWB_PHY_NUMOF_SYM_PHR		18
+
+/* Data Symbol Duration Tdsym in ns */
+#define UWB_PHY_TDSYM_DATA_110K		8205.13
+#define UWB_PHY_TDSYM_DATA_850K		1025.64
+#define UWB_PHY_TDSYM_DATA_6M8		128.21
+
+#define DWT_WORK_QUEUE_STACK_SIZE	512
+
+static struct k_work_q dwt_work_queue;
+static K_THREAD_STACK_DEFINE(dwt_work_queue_stack,
+			     DWT_WORK_QUEUE_STACK_SIZE);
+
+struct dwt_phy_config {
+	u8_t channel;	/* Channel 1, 2, 3, 4, 5, 7 */
+	u8_t dr;	/* Data rate DWT_BR_110K, DWT_BR_850K, DWT_BR_6M8 */
+	u8_t prf;	/* PRF DWT_PRF_16M or DWT_PRF_64M */
+
+	u8_t rx_pac_l;		/* DWT_PAC8..DWT_PAC64 */
+	u8_t rx_shr_code;	/* RX SHR preamble code */
+	u8_t rx_ns_sfd;		/* non-standard SFD */
+	u16_t rx_sfd_to;	/* SFD timeout value (in symbols)
+				 * (tx_shr_nsync + 1 + SFD_length - rx_pac_l)
+				 */
+
+	u8_t tx_shr_code;	/* TX SHR preamble code */
+	u32_t tx_shr_nsync;	/* PLEN index, e.g. DWT_PLEN_64 */
+
+	float t_shr;
+	float t_phr;
+	float t_dsym;
+};
+
+struct dwt_hi_cfg {
+	const char *irq_port;
+	u8_t irq_pin;
+	gpio_dt_flags_t irq_flags;
+	const char *rst_port;
+	u8_t rst_pin;
+	gpio_dt_flags_t rst_flags;
+	const char *spi_port;
+	u8_t spi_cs_pin;
+	const char *spi_cs_port;
+	u32_t spi_freq;
+	u8_t spi_slave;
+};
+
+#define DWT_STATE_TX		0
+#define DWT_STATE_CCA		1
+#define DWT_STATE_RX_DEF_ON	2
+
+struct dwt_context {
+	struct net_if *iface;
+	struct device *irq_gpio;
+	struct device *rst_gpio;
+	struct device *spi;
+	struct spi_cs_control spi_cs;
+	struct spi_config *spi_cfg;
+	struct spi_config spi_cfg_slow;
+	struct spi_config spi_cfg_fast;
+	struct gpio_callback gpio_cb;
+	struct k_sem dev_lock;
+	struct k_sem phy_sem;
+	struct k_work irq_cb_work;
+	struct k_thread thread;
+	struct dwt_phy_config rf_cfg;
+	atomic_t state;
+	bool cca_busy;
+	u16_t sleep_mode;
+	u8_t mac_addr[8];
+};
```

Comment:
btw, why not putting all of these above in the header file as well?

Reply:
ieee802154_dw1000_reg.h contains everything about register. It was more practical for me to have it here.